pytorch_lightning/plugins/training_type/tpu_spawn.py

```diff
@@ -0,0 +1,196 @@
+        data = xm.all_gather(data_tensor)
+        buffer = io.BytesIO(data.cpu().byte().numpy())
+        obj = torch.load(buffer)
+        return obj
+
+    def load_spawn_weights(self, original_model: LightningModule) -> LightningModule:
+        """
+        Load the temp weights saved in the process
+        To recover the trained model from the ddp process we load the saved weights
+        """
+
+        loaded_model = original_model
+
+        if self.is_global_zero:
+            # load weights saved in ddp
+            path = os.path.join(original_model.trainer.default_root_dir, "__temp_weight_distributed_end.ckpt")
+            loaded_model = original_model.__class__.load_from_checkpoint(path)
+
+            # copy loaded weights to old model
+            original_model.load_state_dict(loaded_model.state_dict())
+
+            # remove ddp weights
+            os.remove(path)
+
+        return loaded_model
+
+    def save_spawn_weights(self, model: LightningModule) -> Optional[str]:
+        """
+        Dump a temporary checkpoint after ddp ends to get weights out of the process
+        """
+        if model.trainer.is_global_zero:
+            path = os.path.join(model.trainer.default_root_dir, "__temp_weight_distributed_end.ckpt")
+            model.trainer.save_checkpoint(path)
+            return path
+
+    def reduce_early_stopping_decision(self, should_stop: bool) -> bool:
+        should_stop = torch.tensor(int(should_stop), device=self.lightning_module.device)
+        stop = xm.mesh_reduce('stop_signal', should_stop, sum)
+        rendezvous("pl.EarlyStoppingCallback.stop_distributed_training_check")
+        should_stop = int(stop.item()) == self.world_size
+        return should_stop
+
+    def post_training(self) -> None:
+        # TODO: Check if trainer references can be resolved otherwise
+        model = self.lightning_module
+
+        # restore main state with best weights
+        best_path = self.mp_queue.get()
+        results = self.mp_queue.get()
+        last_path = self.mp_queue.get()
+
+        # transfer back the best path to the trainer
+        if self.lightning_module.trainer.checkpoint_callback is not None:
+            self.lightning_module.trainer.checkpoint_callback.best_model_path = best_path
+        # todo, pass also bets score
+
+        # load last weights
+        if last_path and not self.lightning_module.trainer.testing:
+            ckpt = torch.load(last_path, map_location=lambda storage, loc: storage)
+            model.load_state_dict(ckpt)
+
+        self.lightning_module = model
+
+        # when training completes, load the weights back in main process
+        self.__load_weights_on_main_process()
+
+    def __load_weights_on_main_process(self) -> None:
+        model = self.lightning_module
+
+        # load weights if not interrupted
+        # TODO: check for trainer reference
+        if self.on_colab_kaggle and not model.trainer.testing:
```

Comment:
Could we have a `ColabEnv for TPU` and move some logic there ? In another PR :)

Reply:
not sure I completely understand this... You mean a `ClusterEnvironment` for that? We can certainly look into that.

Reply:
Not sure Colab is a `ClusterEnviromenent`.

I meant more something like `TPURunningEnviromenent` -> Colab, Kaggle_colab, etc...

Best,
T.C